### PR TITLE
Improve Connection Typings

### PR DESCRIPTION
**Pull request by Uzlopak**
This PR extracts the Connection Typings

With the first commit I extract the typings from Connection from index.d.ts into a new Connection.d.ts. 
With the second commit I do the actual improvements.

The order of the function overloads are according to Typescript Do's and Don'ts, so basically by most arity to least arity.
Attributes like pass, user, models etc. were set to readonly. nobody should use them directly for manipulating the internals of mongoose. If there is some reason to modify them, a developer should use the corresponding methods.

Just check the second commit to see relevant changes.

Also added typings tests.

### CODE REVIEW

File: types/Connection.d.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+      readonly models: Readonly<{ [index: string]: Model<any> }>;
+
+      /** Defines or retrieves a model. */
+      model<T, U, TQueryHelpers = {}>(
+            name: string,
+            schema?: Schema<T, U, TQueryHelpers>,
+            collection?: string,
+            options?: CompileModelOptions
+        ): U;
+      model<T>(name: string, schema?: Schema<T>, collection?: string, options?: CompileModelOptions): Model<T>;
```
Comment (Uzlopak) [inline types/Connection.d.ts:127]: Question to all:
shouldnt we do here:
model<T extends AnyObject = AnyObject>

With the current typings we are forcing a dev to type .model(). I think better DX is to make it optional for the dev.

File: types/Connection.d.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+         *
+         * - 0 = disconnected
+         * - 1 = connected
+         * - 2 = connecting
+         * - 3 = disconnecting
+         * - 99 = uninitialized
+         */
+      readonly readyState: ConnectionStates;
+
+      /** Sets the value of the option `key`. */
```
Comment (Uzlopak) [inline types/Connection.d.ts:166]: here we had the information that set is the same as connection.options[key] = value.

The typings do not expose the options attribute, so I removed this remark here and the same for get

File: types/Connection.d.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+         * Returns the [MongoDB driver `MongoClient`](http://mongodb.github.io/node-mongodb-native/3.5/api/MongoClient.html) instance
+         * that this connection uses to talk to MongoDB.
+         */
+      getClient(): mongodb.MongoClient;
+
+      /**
+         * The host name portion of the URI. If multiple hosts, such as a replica set,
+         * this will contain the first host name in the URI
+         */
+      readonly host: string;
```
Comment (Uzlopak) [inline types/Connection.d.ts:105]: please give me feedback regarding the readonly on host, user, pass etc. 
Is it wrong? Is there actually a use case for having them writeable?

If I am wrong, then I would like to fix it before we have regression reports

File: types/Connection.d.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+       */
+      readonly models: Readonly<{ [index: string]: Model<any> }>;
+
+      /** Defines or retrieves a model. */
+      model<T, U, TQueryHelpers = {}>(
+            name: string,
+            schema?: Schema<T, U, TQueryHelpers>,
+            collection?: string,
+            options?: CompileModelOptions
+        ): U;
```
Comment (Uzlopak) [inline types/Connection.d.ts:126]: ReturnType U seems odd. Should it be Model<U> or does U should extend from Model?

Comment (vkarpov15) [reply types/Connection.d.ts:126]: `U` should extend from Model, but I believe we removed that for perf reasons in #10349. `U extends Model` causes perf issues for the TS compiler

File: types/Connection.d.ts (hunk 3 above)
Comment (vkarpov15) [reply types/Connection.d.ts:105]: I think `readonly` should be fine, modifying `host` is a noop.

**Review by vkarpov15 — APPROVED** [review]

**Pull request merged**